src/openloco/interop/interop.hpp

```diff
@@ -130,58 +130,180 @@ namespace openloco::interop
     int32_t call(int32_t address);
     int32_t call(int32_t address, registers& registers);
 
-    template<typename T, uint32_t TAddress>
+    template<typename T, uintptr_t TAddress>
     struct loco_global
     {
-        void operator=(T rhs)
+        typedef T type;
+        typedef type* pointer;
+        typedef type& reference;
+        typedef const type& const_reference;
+
+        operator reference()
         {
-            addr<TAddress, T>() = rhs;
+            return addr<TAddress, T>();
         }
 
-        operator T&()
+        loco_global& operator=(const_reference v)
         {
-            return addr<TAddress, T>();
+            addr<TAddress, T>() = v;
+            return *this;
         }
 
-        T& operator*()
+        loco_global& operator+=(const_reference v)
+        {
+            addr<TAddress, T>() += v;
+            return *this;
+        }
+
+        loco_global& operator|=(const_reference v)
+        {
+            addr<TAddress, T>() |= v;
+            return *this;
+        }
+
+        loco_global& operator&=(const_reference v)
+        {
+            addr<TAddress, T>() &= v;
+            return *this;
+        }
+
+        loco_global& operator-=(const_reference v)
+        {
+            addr<TAddress, T>() -= v;
+            return *this;
+        }
+
+        loco_global& operator++()
+        {
+            addr<TAddress, T>()++;
+            return *this;
+        }
+
+        loco_global operator++(int)
+        {
+            addr<TAddress, T>()++;
+            return *this;
```

Comment:
This is just wrong.

Suppose there's a global like this:
```cpp
loco_global<int32_t, 0x12345678> example;
```
And suppose that the value of the `int32_t` at address `0x12345678` is currently `99`.

Then someone does this in their code:
```cpp
int32_t x = example++;
```
The programmer who writes that code expects `x` to have the value `99`. But `x` will instead have value `100`, because the post-increment operator is currently implemented identically to the pre-increment operator: it increments the value at the `0x12345678` from 99 to 100, and then returns a `loco_global` for `0x12345678`,  the value at the address **after the increment operation**.

The fact that you're returning a copy of the `loco_global` by-value rather than by-reference doesn't actually accomplish anything, because the original `loco_global` and a copied old version of the `loco_global` *are not different* after an arithmetic operation is done, because the arithmetic operation *affects the pointed-to value*. The `loco_global` just stores the address; it knows nothing about the status of *the value at that address*.

Or in other words: this is a different implementation situation for operator overloading than what you'd do for a smart pointer or an iterator, where the increment or decrement operation affects the smartptr/iterator's address/offset itself, and so you can simply return an old copy of the smartptr/iterator.

So either:
- (a) `loco_global::operator++(int)` needs to have return type `T`, and store a temporary `T` before doing the increment, and return that temporary `T`; or
- (b) if (a) is objectionable due to the fact that you're no longer returning a `loco_global` wrapper of some kind and are simply returning a value of type `T` which can't be chained for further operations as a `loco_global` could be, then the post-increment operator should just be explicitly deleted (`void operator++(int) = delete;`) and users of `loco_global` should be therefore forced to use only pre-increment or pre-decrement, to prevent them from making mistakes based on their expectations of how post-increment of a value-wrapper-type like `loco_global` would work.

----

Here's an test you can do to show that the current implementation works incorrectly. Put this at the beginning of `main`:
```cpp
uint32_t before  =   _srand0;   std::printf("%08X\n", before);
uint32_t preinc  = ++_srand0;   std::printf("%08X\n", preinc);
uint32_t postinc =   _srand0++; std::printf("%08X\n", postinc);
uint32_t after   =   _srand0;   std::printf("%08X\n", after);
```
Expected output:
```
00000000
00000001
00000001
00000002
```
Actual output:
```
00000000
00000001
00000002
00000002
```